### ISSUE TITLE
How to check that spark has started without polling?

### ISSUE THREAD

**Issue by tchoulihan** [perwendel/spark#226]
I have a program that needs to load spark a 'get' resource, but only after spark has started

so my current main is:
````{
spark.start();
loadSparkURL();
}````

Yet oftentimes, since I suppose spark uses an asynchronous start, the URL needs to be refreshed. Is there any way to check without polling that URL every so often?

**Comment by amarseillan** [comment]
I don't think there's such thing as a "start" method... Maybe you are using an older version of spark??

Spark initialization is checked/made in every route registration, the private method init() is called every time you want to register a route, so you don't have to check that spark is initialized.
This method is synchronized so don't worry about asynchronism.

the init method:

```java
    private static synchronized void init() {
        if (!initialized) {
            routeMatcher = RouteMatcherFactory.get();
            new Thread(new Runnable() {
                @Override
                public void run() {
                    server = SparkServerFactory.create(hasMultipleHandlers());
                    server.ignite(
                            ipAddress,
                            port,
                            keystoreFile,
                            keystorePassword,
                            truststoreFile,
                            truststorePassword,
                            staticFileFolder,
                            externalStaticFileFolder);
                }
            }).start();
            initialized = true;
        }
```

**Comment by amarseillan** [comment]
What version of spark are you using? I've never seen a star() method exposed

**Comment by tchoulihan** [comment]
the start() is my own method, that contains all the gets and posts. The setting up is probably asynchronous, because if you try to access a resource immediately after you set it up with get(...) it will fail.

**Comment by amarseillan** [comment]
Perfect, now I get your question, I'll check it out

**Comment by amarseillan** [comment]
Yup, now I see it... You are totally right...

I'll check how this should be done!

**Comment by tchoulihan** [comment]
I'd love to see it implemented in some form of event listener, or
listenablefuture.

**Comment by amarseillan** [comment]
Ok, I resolved it by waiting on a lock after starting the server thread. After the server initialization, before joining the threadpool, I notifyAll() on that lock... Best idea I had given the architecture.

Better ideas are accepted =P, i'll send in a PR.

Does it make sense to make this optional?? I don't think so but well...

**Comment by tchoulihan** [comment]
Here's a really good page on [listenable futures](https://code.google.com/p/guava-libraries/wiki/ListenableFutureExplained) . They're pretty much the best way to handle asynchronous tasks like we have here. Either the spark startup, or each get/post could implement the listenablefuture interface. 

Then, spark users that want to run something when its finished, can call
```
Futures.addCallback(spark, new FutureCallback<Spark>() {
  public void onSuccess(Spark spark) {
     // What to do when spark has started
  }
  public void onFailure(Throwable thrown) {
   // Spark couldn't start
  }
});
```

I'd highly suggest against program-locking here, since it would just slow things down and is unecessary.

**Comment by amarseillan** [comment]
I agree that listenable futures is the best way to handle asynchronous
tasks. I don't agree about this being an asynchronous task... If we wanted
to be able to register routes while starting the server to gain time while
starting the server (which doesn't make THAT much sense...) we should
manage states, which jetty does and can be used. But using something that
is not initialized is not acceptable IMO.

Anyway, this never came up because generally no one uses the same java
program to make HTTP request to the same java program acting as a server.

**Comment by tchoulihan** [comment]
Gotcha, I could see the notifyAll() as working, as long as it synchronizes after spark has added all its routes.

In my app, I do need this functionality(I load a spark resource after spark has started, but right now I have to do it with polling), and there appears to be another github issue referencing a need to fix the same problem.

**Comment by amarseillan** [comment]
Sure, I'm not saying it shouldn't be fixed, I'll propose a fix =D

Just pointing that out

**Comment by amarseillan** [comment]
There, I think that PR will fix your code... Anyway, Spark isn't thread safe still, if requests start coming in while you are registering a route, that would make 2 threads to access de _routes_ list at the same time, one iterating it and the other one adding a route, which would crash.

**Comment by amarseillan** [comment]
I actually started having a similar issue with spark. The fact that the web server has started but not all routes are registered is pretty bad, because I use a health check URL to see if my application is up and running... If I'm not able to connect to the aplication then I'll keep trying... but if I'm able to connect to it and get a 404 because my /health-check is not registered, well that's just wrong.

I fixed this cloning spark and implementing the initialization my way.
